leaderboard/competition_ranking_leaderboard.py

```diff
@@ -150,3 +131,55 @@ def __rankings_for_members_having_scores_in(self, leaderboard_name, members, sco
         responses = pipeline.execute()
 
         return [self.__up_rank(response) for response in responses]
+
+    def _members_from_rank_range_internal(
+            self, leaderboard_name, start_rank, end_rank, members_only=False, **options):
+        '''
+        Format ordered members with score as efficiently as possible.
+        '''
+        response = self._range_method(
+            self.redis_connection,
+            leaderboard_name,
+            start_rank,
+            end_rank,
+            withscores=not members_only)
+
+        if members_only or not response:
+            return [{self.MEMBER_KEY: member} for member in response]
+
+        # Find out where the current rank started using the first two ranks
+        current_rank = None
+        current_score = None
+        current_rank_start = 0
+        for index, (member, score) in enumerate(response):
+            if current_score is None:
+                current_rank = self.rank_for_in(leaderboard_name, member)
```

Comment:
This is true, but in a different way, with the current implementation. We first fetch the members then rank them all at the same time. If scores change in between those two queries we could be:
* Including players that are no longer in the range.
* Missing players that are now in the range
* Showing members out of order (although I think it previously resorted in RAM to catch this edge case)

I do see that in this current implementation though it is technically possible that the first two scores and ranks change in the gap causing wrong/duplicate ranks downstream. I'll think a bit more on this point.

Reply:
I think the correct path forward in either case is to use LUA scripts which can have [transaction-like semantics](https://redis.io/commands/eval#atomicity-of-scripts). This would ensure that we have a consistent view of the data while we do more advanced calculations and would be even more performant.

Moving forward on that path would be a huge undertaking though, essentially moving chunks of our python to LUA scripts. Thoughts?

Reply:
A change to using LUA scripts would essentially be a rewrite of most of the core functionality and while it's unclear if there'd be breaking changes, the changes required there feel like that'd warrant a major version bump. So I'm a big 👎 at the moment for these changes outside of LUA, but the smaller readability, variable name, and helper method changes seem worthy of a separate PR.